# Example Notebook
This notebook demonstrates some of the ways we can use a notebook for data science.

There are lots of shortcuts, so check out "Help"->"Keyboard shortcuts" for a list. Here are some of the ones I use routinely:

  * ctrl+enter (Windows) or cmd+enter (macOS) -- runs the current cell
  * shift+enter -- runs the current cell then moves to the next cell
  * alt+enter -- runs the current cell, creates a new cell below, then moves to that cell
  * esc -- pressing this will put you into command mode for the selected shell; once there you can press...
    - a -- adds a cell above, then moves to that cell in command mode
    - b -- adds a cell below, then moves to that cell in command mode
    - m -- convert the cell to a Markdown cell
    - y -- convert the cell to a code cell
    - enter -- puts you into edit mode for the selected cell
    - up / down arrows to move between cells
  * in edit mode:
    - wihout anything selected, ctrl+x | cmd+x will cut the current line
    - ctrl+/ | cmd+/ will toggle comments for the selected lines
    - while typing a function name, hit shift+tab a few times to bring up the documentation in various ways

$x = \sqrt{a^5\times\frac{c}{d_1^4}}$

In [ ]:
sum()

In [2]:
x = 10

In [3]:

x = x**2
x

100

In [4]:
y = 50
y
x = 5
x

5

In [5]:
%%bash
# What's %%bash does?
pwd
ls

/mnt/c/Users/hfeild/Documents/classes/2025-sp/csc440/code/notebooks
03-example-notebook.ipynb
04-web-scraping.ipynb
demo.ipynb
rand-count-plot.ipynb


# Question 1
What does the following do?

In [1]:
import sys
sys.path.append('../scripts')

In [6]:
sys.path

['/mnt/c/Users/hfeild/Documents/classes/2025-sp/csc440/code/notebooks',
 '/mnt/c/Users/hfeild/Documents/classes/2025-sp/csc440/code/notebooks/PYTHONPATH',
 '//home/hfeild/research/thesis/scripts',
 '/usr/lib/python38.zip',
 '/usr/lib/python3.8',
 '/usr/lib/python3.8/lib-dynload',
 '',
 '/home/hfeild/.local/lib/python3.8/site-packages',
 '/usr/local/lib/python3.8/dist-packages',
 '/usr/lib/python3/dist-packages',
 '/usr/lib/python3/dist-packages/IPython/extensions',
 '/home/hfeild/.ipython',
 '../scripts']

In [7]:
import count_bots

In [9]:
# HW 6

import count_bots
import apache2cols

with open('../data/apache-logs/hank.feild.org-access.log', 'r') as file:
    apacheData = [apache2cols.splitIntoColumns(line) for line in file]
    counts = count_bots.countBots(apacheData)
    
counts

defaultdict(int,
            {'Mozilla/5.0 (compatible; YandexBot/3.0; +http://yandex.com/bots)': 3,
             'Mozilla/5.0 (compatible; YandexBot/3.0; +http://yandex.com/bots) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/81.0.4044.268': 13,
             'Mozilla/5.0 (compatible; Yellowbrandprotectionbot/1.0; +https://www.yellowbp.com/bot.html)': 4,
             'Mozilla/5.0 (Linux; Android 6.0.1; Nexus 5X Build/MMB29P) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/107.0.5304.115 Mobile Safari/537.36 (compatible; Googlebot/2.1; +http://www.google.com/bot.html)': 14,
             'Mozilla/5.0 (Linux; Android 6.0.1; Nexus 5X Build/MMB29P) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/108.0.5359.130 Mobile Safari/537.36 (compatible; Googlebot/2.1; +http://www.google.com/bot.html)': 2,
             'Mozilla/5.0 AppleWebKit/537.36 (KHTML, like Gecko; compatible; bingbot/2.0; +http://www.bing.com/bingbot.htm) Chrome/103.0.5060.134 Safari/537.36': 8,
             'Mozilla/5.0 (Linux; Andr

# Aapache log analysis

How many events are in all 52 logs?

In [10]:
%%bash
cd ../data/apache-logs
zcat -f * | wc -l

206488


We can capture the output of the bash environment into a Python variable using `--out`:

In [11]:
%%bash --out eventCount
cd ../data/apache-logs
zcat -f * | wc -l

In [12]:
eventCount

'206488\n'

We can parse it as an int to treat it numerically. Here's we're getting the average event count per log.

In [13]:
int(eventCount) / 52

3970.923076923077

# Python practice

Iterate over an interval, e.g., 
print numbers 4 to 103
